sacrebleu/metrics/ter.py

```diff
@@ -0,0 +1,592 @@
+# (extreme) corner cases, the output could differ.
+#
+# Tercom original implementation:
+#
+# https://github.com/jhclark/tercom
+#
+# Caching in the edit distance is based partly on the PyTer package by Hiroyuki
+# Tanaka (MIT license).
+#
+# https://github.com/aflc/pyter
+
+_COST_INS = 1
+_COST_DEL = 1
+_COST_SUB = 1
+
+# Tercom-inspired limits
+_MAX_SHIFT_SIZE = 10
+_MAX_SHIFT_DIST = 50
+_BEAM_WIDTH = 20
+
+# Our own limits
+_MAX_CACHE_SIZE = 10000
+_MAX_SHIFT_CANDIDATES = 1000
+
+_OP_INS = 'i'
+_OP_DEL = 'd'
+_OP_NOP = ' '
+_OP_SUB = 's'
+_OP_UNDEF = 'x'
+
+
+class TERScore(BaseScore):
+    def __init__(self, num_edits, ref_length):
+        score = num_edits / ref_length if ref_length > 0 else 1
+        super().__init__(score)
+
+        self.num_edits = num_edits
+        self.ref_length = ref_length
+        self.prefix = 'TER'
+
+    def format(self, width=2, score_only=False, signature=''):
+        if score_only:
+            return '{0:.{1}f}'.format(self.score, width)
+
+        prefix = "{}+{}".format(self.prefix, signature) if signature else self.prefix
+        return '{pr} = {sc:.{w}f}'.format(pr=prefix, sc=self.score, w=width)
+
+
+class TERSignature(Signature):
+    def __init__(self, args, tokenizer):
+        super().__init__(args)
+        self.info.update({"tok": tokenizer.signature()})
+
+
+class TER:
+    def __init__(self, args, tokenizer=None):
+        if tokenizer is None:
+            tokenizer = TOKENIZERS['tercom']()
+
+        self.tokenizer = tokenizer
+        self.signature = TERSignature(args, tokenizer)
+
+    def corpus_score(self, sys_stream: Union[str, Iterable[str]],
+                     ref_streams: Union[str, List[Iterable[str]]]) -> TERScore:
+        # Add some robustness to the input arguments
+        if isinstance(sys_stream, str):
+            sys_stream = [sys_stream]
+
+        if isinstance(ref_streams, str):
+            ref_streams = [[ref_streams]]
+
+        fhs = [sys_stream] + ref_streams
+
+        total_edits = 0
+        sum_ref_lengths = 0
+
+        for lines in zip_longest(*fhs):
+            if None in lines:
+                raise EOFError("Source and reference streams have different lengths!")
+            hypo, *refs = lines
+
+            words_hyp = self.tokenizer(hypo).split()
+
+            best_num_edits = None
+            ref_lengths = 0
+
+            for ref in refs:
+                words_ref = self.tokenizer(ref).split()
+                num_edits, ref_len = translation_edit_rate(words_hyp, words_ref)
+                ref_lengths += ref_len
+                if best_num_edits is None or num_edits < best_num_edits:
+                    best_num_edits = num_edits
+
+            total_edits += best_num_edits
+            sum_ref_lengths += (ref_lengths / len(refs))
+
+        return TERScore(total_edits, sum_ref_lengths)
+
+    def sentence_score(self, hypothesis: str, references: List[str]) \
+            -> TERScore:
```

Comment:
We (and you as well :)) already have longer lines than this so I would lean towards not using the continuation marker and move `-> TERScore` above. (In remaining codes as well).

Reply:
Yes, good point. Most of the code was taken from an internal code base where we try to follow the Google style guide. I removed the backslashes, hopefully I didn't miss any.

Reply:
I also converted the original Google-style docstrings into RST.